src/object/obj_tx.c

```diff
@@ -0,0 +1,1063 @@
+	.hop_free	= dc_tx_free,
+};
+
+static void
+dc_tx_decref(struct dc_tx *tx)
+{
+	daos_hhash_link_putref(&tx->tx_hlink);
+}
+
+static struct dc_tx *
+dc_tx_hdl2ptr(daos_handle_t th)
+{
+	struct d_hlink	*hlink;
+
+	hlink = daos_hhash_link_lookup(th.cookie);
+	if (hlink == NULL)
+		return NULL;
+
+	return container_of(hlink, struct dc_tx, tx_hlink);
+}
+
+static daos_handle_t
+dc_tx_ptr2hdl(struct dc_tx *tx)
+{
+	daos_handle_t	th;
+
+	daos_hhash_link_key(&tx->tx_hlink, &th.cookie);
+
+	return th;
+}
+
+static void
+dc_tx_hdl_link(struct dc_tx *tx)
+{
+	daos_hhash_link_insert(&tx->tx_hlink, DAOS_HTYPE_TX);
+}
+
+static void
+dc_tx_hdl_unlink(struct dc_tx *tx)
+{
+	daos_hhash_link_delete(&tx->tx_hlink);
+}
+
+static int
+dc_tx_alloc(daos_handle_t coh, daos_epoch_t epoch, uint32_t flags,
+	    struct dc_tx **ptx)
+{
+	daos_handle_t	 ph;
+	struct dc_tx	*tx;
+
+	if (daos_handle_is_inval(coh))
+		return -DER_NO_HDL;
+
+	ph = dc_cont_hdl2pool_hdl(coh);
+	D_ASSERT(!daos_handle_is_inval(ph));
+
+	D_ALLOC_PTR(tx);
+	if (tx == NULL)
+		return -DER_NOMEM;
+
+	tx->tx_pool = dc_hdl2pool(ph);
+	D_ASSERT(tx->tx_pool != NULL);
+
+	D_INIT_LIST_HEAD(&tx->tx_sub_reqs);
+	daos_dti_gen(&tx->tx_id, false);
+	/* If @epoch is zero, then tx_epoch will be generated
+	 * by the first accessed server.
+	 */
+	tx->tx_epoch = epoch;
+	tx->tx_coh = coh;
+	tx->tx_flags = flags;
+	tx->tx_status = TX_OPEN;
+	daos_hhash_hlink_init(&tx->tx_hlink, &tx_h_ops);
+	dc_tx_hdl_link(tx);
+
+	*ptx = tx;
+
+	return 0;
+}
+
+static void
+dc_tx_cleanup(struct dc_tx *tx)
+{
+	struct dc_tx_sub_req	*dtsr;
+	int			 i;
+
+	while ((dtsr = d_list_pop_entry(&tx->tx_sub_reqs,
+					struct dc_tx_sub_req,
+					dtsr_link)) != NULL) {
+		switch (dtsr->dtsr_opc) {
+		case DAOS_OBJ_RPC_UPDATE:
+			if (dtsr->dtsr_zero_copy)
+				break;
+
+			for (i = 0; i < dtsr->dtsr_nr; i++) {
+				daos_iov_free(&dtsr->dtsr_iods[i].iod_name);
+				D_FREE(dtsr->dtsr_iods[i].iod_recxs);
+				daos_sgl_fini(&dtsr->dtsr_sgls[i], true);
+			}
+
+			D_FREE(dtsr->dtsr_iods);
+			D_FREE(dtsr->dtsr_sgls);
+			daos_iov_free(dtsr->dtsr_dkey);
+			break;
+		case DAOS_OBJ_RPC_PUNCH:
+			break;
+		case DAOS_OBJ_RPC_PUNCH_DKEYS:
+			if (!dtsr->dtsr_zero_copy)
+				daos_iov_free(dtsr->dtsr_dkey);
+			break;
+		case DAOS_OBJ_RPC_PUNCH_AKEYS:
+			if (!dtsr->dtsr_zero_copy) {
+				for (i = 0; i < dtsr->dtsr_nr; i++)
+					daos_iov_free(&dtsr->dtsr_akeys[i]);
+
+				D_FREE(dtsr->dtsr_akeys);
+				daos_iov_free(dtsr->dtsr_dkey);
+			}
+			break;
+		default:
+			D_ASSERT(0);
+		}
+
+		obj_decref(dtsr->dtsr_obj);
+		D_FREE(dtsr);
+	}
+
+	tx->tx_sub_count = 0;
+}
+
+static int
+dc_tx_restart_internal(struct dc_tx *tx, daos_epoch_t epoch)
+{
+	dc_tx_cleanup(tx);
+	if (epoch != 0 && epoch <= tx->tx_epoch) {
+		D_ERROR("New epoch %lu is not larger than TX old one %lu\n",
+			epoch, tx->tx_epoch);
+		return -DER_INVAL;
+	}
+
+	tx->tx_status = TX_OPEN;
+	tx->tx_epoch = epoch;
+
+	return 0;
+}
+
+int
+dc_tx_restart(daos_handle_t th, daos_epoch_t epoch)
```

Comment:
I am not sure I follow the idea. Some of you does not agree to export TX API, so we offer an daos internal function dc_tx_restart() that will be called by other object layer functions or RPC callback when need to restart related TX. For example, if the MVCC on detect some conflict, it will return -DER_AGAIN to the client plus a suggested epoch for retry. Under such case, the client side RPC callback logic will call dc_tx_restart() with such replied epoch.

Reply:
The idea is simply that the public restart API method should not take an epoch parameter.

Reply:
The new TX restart API does not take "epoch" parameter.

Reply:
Yes, the new public restart method looks fine. I was replying to our previous messages.